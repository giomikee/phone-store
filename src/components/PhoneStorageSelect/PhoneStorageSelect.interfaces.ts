import type { StorageOption } from '../../interfaces';

export interface Props {
    onSelectedStorage: (storage: string) => void;
    storageOptions?: StorageOption[];
    selectedStorage?: string;
}