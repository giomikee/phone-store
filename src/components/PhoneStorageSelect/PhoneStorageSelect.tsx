import { FormControl, InputLabel, MenuItem, Select, Skeleton } from '@mui/material';
import type { Props } from './PhoneStorageSelect.interfaces';
import { formatPrice } from '../../utils';

const PhoneStorageSelect = ({ selectedStorage, storageOptions, onSelectedStorage }: Props) => {
    if (!storageOptions) {
        return <Skeleton height={32} />
    }

    return (
        <FormControl fullWidth variant='filled'>
            <InputLabel 
                htmlFor='storage-select'
                id='storage-select-label'
            >
                Almacenamiento
            </InputLabel>

            <Select
                color={storageOptions ? 'success' : undefined}
                id='storage-select'
                data-testid='storage-select'
                name='storage-select'
                label='Almacenamiento'
                labelId='storage-select-label'
                value={selectedStorage ?? ''}
                onChange={(e) => onSelectedStorage(e.target.value)}
            >
                {
                    storageOptions.map(({ capacity, price }) => (
                        <MenuItem 
                            key={capacity} 
                            value={capacity}
                        >
                            {capacity} ({ formatPrice(price) })
                        </MenuItem>
                    ))
                }
            </Select>
        </FormControl>
    )
}

export default PhoneStorageSelect