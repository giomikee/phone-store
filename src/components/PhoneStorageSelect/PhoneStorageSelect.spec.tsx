import { describe, expect, it, vi } from 'vitest';
import { fireEvent, render, screen } from '@testing-library/react';
import '@testing-library/jest-dom/vitest';
import PhoneStorageSelect from './PhoneStorageSelect';

describe('Given PhoneStorageSelect component', () => { 
    const defaultProps = {
        storageOptions: [
            {
                capacity: 'capacity',
                price: 59
            }
        ],
        onSelectedStorage: vi.fn(),
    }

    describe.each([
        defaultProps.storageOptions,
        undefined
    ])('When component is rendered with storageOptions=%s', (storageOptions) => {
        it('Then it should match the snapshot', () => {
            const { container } = render(<PhoneStorageSelect {...defaultProps} storageOptions={storageOptions} />);
            expect(container).toMatchSnapshot();
        });
    });

    describe('When selected storage is changed', () => {
        it('Then onSelectedStorage should be called with the selected color', () => {
            render(<PhoneStorageSelect {...defaultProps} />);

            const select = screen.getByTestId('storage-select').querySelector('input');

            expect(select).not.toBeNull();
            
            fireEvent.change(select!, { target: { value: 'capacity' } });
            
            expect(defaultProps.onSelectedStorage).toHaveBeenCalledWith('capacity');
        });
    });
});