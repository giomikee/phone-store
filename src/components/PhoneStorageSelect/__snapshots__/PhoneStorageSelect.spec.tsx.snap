// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Given PhoneStorageSelect component > When component is rendered with storageOptions={ capacity: 'capacity', price: 59 } > Then it should match the snapshot 1`] = `
<div>
  <div
    class="MuiFormControl-root MuiFormControl-fullWidth css-ytlejw-MuiFormControl-root"
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-filled MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-filled css-1u43uf3-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink="false"
      for="storage-select"
      id="storage-select-label"
    >
      Almacenamiento
    </label>
    <div
      class="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiInputBase-colorSuccess MuiInputBase-formControl MuiSelect-root css-zz817j-MuiInputBase-root-MuiFilledInput-root-MuiSelect-root"
      data-testid="storage-select"
    >
      <div
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-labelledby="storage-select-label storage-select"
        class="MuiSelect-select MuiSelect-filled MuiInputBase-input MuiFilledInput-input css-1y2qbti-MuiNativeSelect-root-MuiSelect-select-MuiInputBase-input-MuiFilledInput-input"
        id="storage-select"
        role="combobox"
        tabindex="0"
      >
        <span
          aria-hidden="true"
          class="notranslate"
        >
          â€‹
        </span>
      </div>
      <input
        aria-hidden="true"
        aria-invalid="false"
        class="MuiSelect-nativeInput css-j0riat-MuiSelect-nativeInput"
        name="storage-select"
        tabindex="-1"
        value=""
      />
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconFilled css-1rfqz7c-MuiSvgIcon-root-MuiNativeSelect-root-MuiSelect-icon"
        data-testid="ArrowDropDownIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
    </div>
  </div>
</div>
`;

exports[`Given PhoneStorageSelect component > When component is rendered with storageOptions=undefined > Then it should match the snapshot 1`] = `
<div>
  <span
    class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-kkcbhd-MuiSkeleton-root"
    style="height: 32px;"
  />
</div>
`;
